Adding belongsTo relation fixes issue

ValenberghsSven opened the pull request:
PR om jenkins te triggeren

Bij verwijderen van een piece op agendaitem/documents route en daarna naar subcase/documents gaan geeft een fout, omdat de submission-acitivity die aan dat piece hangt die piece nog kent, terwijl die al is verwijderd
meer specifiek, dit gebeurt enkel als het piece op agendaitem vb versie BIS was en we die BIS hebben vervangen omdat er al een goedgekeurde agenda is met het origineel document.
Bij deleten van de BIS gaan we dan het piece deleten, en na de piece een call doen naar de service om het origineel document (dat niet meer rechtstreeks aan het agendaitem hing) terug te hangen aan het agendaitem.

Mijn mening:
Tussen piece en submission activity is piece de parent (belongsTo ) en submission de child (hasMany pieces)
Echter bij creatie van data gaan we op child kant de relatie saven, dus de array van pieces setten en saven.

Piece model had geen weet van submission-activity relaties, daardoor ging ember de child kant niet op de hoogte brengen bij verwijderen van de parent.
Aan de child kant een inverse:null helpt niet, we verwachten van ember dat belongsTo/hasMany automatisch op de hoogte worden gebracht van changes, maar we linken de modellen niet correct.

Bij deleten van piece is het piece niet helemaal weg, de store kent deze piece nog 'in flight - deleted' dus je kan deze nog bekijken. Dus submission-activity heeft alle pieces nog, ook de deleted pieces


## EDIT 29/03/2021:
https://github.com/kanselarij-vlaanderen/kaleidos-project/tree/bugfix/submission-activity-inludes-pieces

Op de falende PATCH call op te lossen heb ik de relatie ook in de lisp files gezet.
De reden dat serialize=false niet werkt is omdat onze application serializer dit niet opvangt.
De serializer kijkt enkel naar die waarde bij hasMany relaties, en niet bij belongsTo relaties.

We zouden in onze serializer de `serializeBelongsTo` method kunnen overschrijven en die value afchecken om dit probleem te vermijden.

ValenberghsSven committed:
Adding belongsTo relation fixes issue

ValenberghsSven committed:
piece - submission relation try inverse null

ValenberghsSven committed:
cypress remove page reload and test again

ValenberghsSven commented:
enkel serialize: false heeft side-effecten (1e build in jenkins failed)

met serialize: false & inverse: null erbij geen side-effecten (2e build in jenkins succes)

(3e build in jenkins zonder de page reload workaround in agendaitem-pieces)

MikiDi commented:
Hopelijk opgelost met https://github.com/kanselarij-vlaanderen/kaleidos-frontend/commit/6ca9007a887c25b29b120b6c0d43c828636867d8 .
Ember-data heeft de relatie-definitie aan de andere kant (`piece`-`submissionActivity`) nodig om op het moment van verwijderen te kunnen weten uit welke relatie-arrays het verwijderde object moet weggehaald worden (het object zelf blijft immers altijd in de store, in een `isDeleted` state).
Omdate ED by default uitgaat van serializatie langs de `belongsTo`-kant, moeten we wel expliciet aan de `hasMany`-kant `serialize: true` specifiëren nu de relatie (in ember-data dan toch) in beide richtingen bestaat.
Nu deze configuratie er staat, en wetende dat deze relatie niet in mu-cl-resources gedefinieerd is, vroeg ik me af hoe ED nu kan weten dat hij voor een `piece.submissionActivity`-opvraging nooit een request naar `pieces/123/submissionActivity` mag sturen (dat zou een error geven). Hiervoor steunt hij op de relationship `links` in de JSONAPI-payload waarmee het piece zelf in eerste instantie opgehaald werd.

MikiDi commented:
@ValenberghsSven geeft aan dat 6ca9007 als resultaat geeft dat ember-data soms toch de (in backend onbestaande) relatie `piece-submissionActivity` gaat proberen serizalizeren.
![Clipboard - March 12, 2021 10 09 AM](https://user-images.githubusercontent.com/9383603/110923450-34ec6880-8321-11eb-9c7b-651be5634faf.png)
Dan toch `serialize: false`? @ValenberghsSven hierboven gaf je aan "enkel serialize: false heeft side-effecten (1e build in jenkins failed)". Weet je nog wat dat dan juist inhield?

ValenberghsSven committed:
Merge branch 'development' of https://github.com/kanselarij-vlaanderen/kaleidos-frontend into bugfix/submission-activity-inludes-pieces

ValenberghsSven committed:
remove duplicate

ValenberghsSven commented:
Ik heb deze branch geupdate met dev, normaal zijn mijn changes dan weg.
Ik heb lokaal getest met serialize: false op development en het effect bleef hetzelfde.
De patch call in de cancel-editing-document.spec.js blijft falen.

Heb die even gepusht hier om te kijken of er dan andere testen falen

De relatie ligt nu in frontend, maar in de document.lisp heeft piece nog geen submission-activity.
De vraag is nu, waarom zit er in die PATCH die submission-activity. 
Als je de pagina refresht en dan dezelfde actie doet, dan zit dat er niet in. (omdat de test sequentieel alles aanmaakt, heeft de store meer relaties geladen dan wanneer je de pagina vernieuwd)
Als je met ember plugin de submission-activity wil ophalen krijg je null, dus de relatie ligt er nu, maar de data is toch niet correct gekoppeld

ValenberghsSven commented:
@MikiDi De reden dat serialize:false op zijn alleen niet lukte was volgens mij omdat we nu in de frontend de modellen hebben gelinkt, maar data correct linken gaat niet aangezien de relatie niet in de lisp files zit.
Dat is ook te merken aan het feit dat zelfs als hangt de piece aan een submission-activity, als je in de store data van piece die gaat getten krijg je null
Voor de volledigheid zullen we volgens mij de lisp file moeten aanpassen en de relatie erin zetten en ervoor zorgen dat frontend die nooit saved

ValenberghsSven commented:
Ondanks dat serialize:false hierbij instaat, faalt die patch op piece nog steeds.
Die patch in cancel-editing.spec bevat submission-activity terwijl dat helemaal niet zou mogen.
Maar dat lijkt enkel voor te komen met meerdere keren achter elkaar edit/cancel te doen.
Ik zou eens tijd moeten steken in bepalen wanneer dit juist voorvalt, want dit mag niet voorvallen op productie natuurlijk

MikiDi commented:
Als er bij iets dat met `serialize: false` gemarkeerd is, weldegelijk nog geserializeerd wordt, dan kijken we eerder naar een bug in ember-data zou ik denken. Deze nog eens met `development` syncen om helemaal zeker te zijn?

ValenberghsSven committed:
revert serialize false

ValenberghsSven committed:
Merge branch 'development' of https://github.com/kanselarij-vlaanderen/kaleidos-frontend into bugfix/submission-activity-inludes-pieces

erikap committed:
Merge branch 'development' into bugfix/submission-activity-inludes-pieces

erikap committed:
Override serializeBelongsTo with option to disable serialization

Similar to the construction used for hasMany relations, definition of
a belongsTo relation also supports a 'serialize' option. If set to
false, the relation will not be serialized. Otherwise fallback to
Ember Data's default behaviour which will always serialize relations
on the belongsTo-side

erikap committed:
Disable serialization of submission-activity on piece model

Since the inverse relation from piece to submission-activity is not
defined in the backend, we never want to serialize this relations in
requests to the backend.

erikap committed:
Convert application serializer to native ES6 class

erikap commented:
Verder bouwend op de "EDIT 29/03/2021" in de originele description van de PR: de inverse relatie van piece naar submission-activity is expliciet vermeden in de backend om door de huidige beperkingen van mu-cl-resources geen custom subpredicate te moeten maken van `prov:generated` voor iedere relatie van piece naar een activity. Besproken met @madnificent , en eens de inheritance-branch landt in mu-cl-resources geldt deze beperking niet langer.

Tot dan lijkt het inderdaad best een `serialize` option te voorzien op een `belongsTo` relation op dezelfde manier zoals we dat al doen voor `hasMany`. Aangezien er geen gelijkaardige method is voor `belongsTo` zoals `shouldSerializeHasMany`  is de method `serializeBelongsTo` van de application serializer overschreven.

Op termijn zullen deze `serialize: false` options dus kunnen verwijderd worden uit de frontend.

erikap committed:
Check if belongsTo serialization option is set to false. Otherwise
fallback to default
